File: legacy/quantum-batch/src/main/java/com/quantum/batch/service/KisTokenService.java
package com.quantum.batch.service;

import com.quantum.core.domain.model.kis.KisToken;
import com.quantum.core.domain.port.repository.KisTokenRepository;
import com.quantum.kis.service.KisRateLimiter;
import com.quantum.kis.service.KisTokenProvider;
import com.quantum.kis.client.KisFeignClient;
import com.quantum.kis.model.KisTokenResponse;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

/**
 * KIS API 토큰 관리 서비스
 * 1분당 1회 제한 대응을 위한 DB 기반 토큰 관리
 * 
 * 다른 배치 프로세서에서도 공통으로 사용할 수 있는 토큰 관리 서비스입니다.
 * 필요시 다른 모듈(quantum-api 등)에서도 유사한 서비스를 만들어 사용하세요.
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class KisTokenService implements KisTokenProvider {

    private final KisFeignClient kisFeignClient;
    private final KisTokenRepository kisTokenRepository;
    private final KisRateLimiter kisRateLimiter;

    @Value("${kis.api.app-key}")
    private String appKey;

    @Value("${kis.api.app-secret}")
    private String appSecret;

    /**
     * 유효한 액세스 토큰 조회 (DB 우선, 없으면 API 호출)
     * 
     * 이 메서드는 thread-safe하며 동시 호출시에도 안전합니다.
     * KIS API의 1분당 1회 제한을 자동으로 준수합니다.
     * 
     * @return 유효한 액세스 토큰, 실패시 null
     */
    public String getValidAccessToken() {
        try {
            // 1. DB에서 활성화된 유효한 토큰 조회
            Optional<KisToken> existingToken = kisTokenRepository.findActiveValidToken();
            
            if (existingToken.isPresent() && !existingToken.get().isExpired()) {
                log.debug("KIS Token Service: Using existing valid token from DB");
                return existingToken.get().getAccessToken();
            }
            
            // 2. 기존 토큰이 없거나 만료된 경우 API로 새 토큰 발급
            log.info("KIS Token Service: Requesting new access token from KIS API");
            KisTokenResponse tokenResponse = requestNewAccessToken();
            
            if (tokenResponse == null || tokenResponse.accessToken() == null) {
                log.error("KIS Token Service: Failed to get new access token from API");
                return null;
            }
            
            // 3. 기존 토큰들 비활성화 후 새 토큰 저장
            kisTokenRepository.deactivateAllTokens();
            
            KisToken newToken = KisToken.fromKisResponse(
                tokenResponse.tokenType(),
                tokenResponse.accessToken(),
                tokenResponse.accessTokenExpired(),
                tokenResponse.expiresIn()
            );
            
            KisToken savedToken = kisTokenRepository.save(newToken);
            log.info("KIS Token Service: New access token saved to DB, expires at: {}", 
                savedToken.getExpiresAt());
            
            return savedToken.getAccessToken();
            
        } catch (Exception e) {
            log.error("KIS Token Service: Error getting valid access token", e);
            return null;
        }
    }
    
    /**
     * KIS API에서 새로운 액세스 토큰 요청
     * 
     * @return KIS 토큰 응답, 실패시 null
     */
    public KisTokenResponse requestNewAccessToken() {
        try {
            Map<String, String> requestBody = new HashMap<>();
            requestBody.put("grant_type", "client_credentials");
            requestBody.put("appkey", appKey);
            requestBody.put("appsecret", appSecret);
            
            KisTokenResponse response = kisFeignClient.getAccessToken(requestBody);
            log.debug("KIS Token Service: Successfully requested new token from API");
            
            return response;
            
        } catch (Exception e) {
            log.error("KIS Token Service: Error requesting new access token from API", e);
            return null;
        }
    }
    
    /**
     * 현재 저장된 토큰 정보 조회 (디버깅/모니터링 용)
     * 
     * @return 활성화된 토큰 정보, 없으면 Optional.empty()
     */
    public Optional<KisToken> getCurrentToken() {
        try {
            return kisTokenRepository.findActiveValidToken();
        } catch (Exception e) {
            log.error("KIS Token Service: Error getting current token info", e);
            return Optional.empty();
        }
    }
    
    /**
     * 모든 토큰 강제 만료 (토큰 재발급 강제 실행용)
     * 
     * 주로 테스트나 운영상 필요에 의해 토큰을 강제로 갱신할 때 사용합니다.
     */
    public void forceTokenRefresh() {
        try {
            log.info("KIS Token Service: Force refreshing all tokens");
            kisTokenRepository.deactivateAllTokens();
            
            // 만료된 토큰들 정리
            kisTokenRepository.deleteExpiredTokens();
            
        } catch (Exception e) {
            log.error("KIS Token Service: Error during force token refresh", e);
        }
    }
    
    /**
     * 토큰 상태 체크 (헬스체크용)
     * 
     * @return 토큰이 유효하면 true, 아니면 false
     */
    public boolean isTokenValid() {
        try {
            Optional<KisToken> token = kisTokenRepository.findActiveValidToken();
            return token.isPresent() && !token.get().isExpired();
        } catch (Exception e) {
            log.error("KIS Token Service: Error checking token validity", e);
            return false;
        }
    }
    
    /**
     * API 호출 전 Rate Limiting 체크
     * KIS API 공식 제한: 실전투자 20건/초, 모의투자 2건/초
     * 
     * @param apiEndpoint API 엔드포인트 구분자
     * @return 호출 가능하면 true, 제한에 걸리면 false
     */
    public boolean checkRateLimit(String apiEndpoint) {
        return kisRateLimiter.tryAcquire(apiEndpoint, false);
    }
    
    /**
     * 강제 대기 후 API 호출 (재시도 포함)
     * 
     * @param apiEndpoint API 엔드포인트 구분자
     * @return 호출 가능하면 true, 최종 실패시 false
     */
    public boolean acquireWithWait(String apiEndpoint) {
        return kisRateLimiter.acquireWithWait(apiEndpoint, false);
    }
    
    /**
     * Rate Limiter 상태 조회 (모니터링용)
     * 
     * @param apiEndpoint API 엔드포인트 구분자
     * @return Rate Limit 상태 정보
     */
    public KisRateLimiter.RateLimitStatus getRateLimitStatus(String apiEndpoint) {
        return kisRateLimiter.getStatus(apiEndpoint);
    }
}

File: quantum-web-api/src/main/kotlin/com/quantum/kis/application/service/KisTokenService.kt
package com.quantum.kis.application.service

import com.quantum.kis.domain.*
import com.quantum.kis.infrastructure.client.KisApiClient
import com.quantum.kis.infrastructure.client.KisApiException
import com.quantum.kis.infrastructure.repository.KisAccountRepository
import com.quantum.kis.infrastructure.repository.KisTokenRepository
import com.quantum.kis.application.dto.KisAccountRequest
import com.quantum.kis.application.dto.KisTokenDto
import kotlinx.coroutines.runBlocking
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.util.concurrent.TimeUnit

/**
 * KIS 토큰 관리 서비스
 * 
 * 하이브리드 토큰 아키텍처의 핵심 서비스
 * 토큰 발급, 갱신, 자동 관리 담당
 */
@Service
@Transactional
class KisTokenService(
    private val kisAccountRepository: KisAccountRepository,
    private val kisTokenRepository: KisTokenRepository,
    private val kisApiClient: KisApiClient
) {
    
    private val logger = LoggerFactory.getLogger(this::class.java)
    
    /**
     * 새 토큰 발급
     * 
     * @param userId 사용자 ID
     * @param request 계정 요청 정보
     * @return 발급된 토큰 정보
     */
    suspend fun issueToken(userId: Long, request: KisAccountRequest): KisTokenDto {
        logger.info("Issuing new KIS token for user: $userId, environment: ${request.environment}")
        
        try {
            // 1. 계정 정보 저장 또는 업데이트
            val kisAccount = saveOrUpdateKisAccount(userId, request)
            
            // 2. 기존 활성 토큰 비활성화
            deactivateExistingTokens(userId, request.environment)
            
            // 3. KIS API로 토큰 발급 요청
            val tokenResponse = kisApiClient.getAccessToken(
                appKey = request.appKey,
                appSecret = request.appSecret,
                environment = request.environment
            )
            
            // 4. 토큰 DB 저장
            val kisToken = KisToken.create(
                userId = userId,
                kisAccountId = kisAccount.id,
                accessToken = tokenResponse.access_token,
                environment = request.environment,
                expiresAt = tokenResponse.getExpiresAt()
            )
            
            val savedToken = kisTokenRepository.save(kisToken)
            
            // 5. 계정에 토큰 발급 시간 기록
            kisAccount.markTokenIssued()
            kisAccountRepository.save(kisAccount)
            
            logger.info("Successfully issued KIS token for user: $userId")
            return KisTokenDto.from(savedToken)
            
        } catch (exception: Exception) {
            logger.error("Failed to issue KIS token for user: $userId", exception)
            throw KisTokenException("토큰 발급 실패: ${exception.message}", exception)
        }
    }
    
    /**
     * 토큰 갱신
     * 
     * @param userId 사용자 ID
     * @param environment KIS 환경
     * @return 갱신된 토큰 정보
     */
    suspend fun refreshToken(userId: Long, environment: KisEnvironment): KisTokenDto {
        logger.info("Refreshing KIS token for user: $userId, environment: ${environment.displayName}")
        
        try {
            // 1. 기존 계정 정보 조회
            val kisAccount = kisAccountRepository.findFirstByUserIdAndEnvironmentAndIsActiveTrueOrderByCreatedAtAsc(
                userId, environment
            ).orElseThrow { 
                KisAccountNotFoundException("활성화된 KIS 계정이 없습니다: userId=$userId, environment=$environment") 
            }
            
            // 2. 기존 토큰 갱신 시작 표시
            val existingToken = kisTokenRepository.findFirstByUserIdAndEnvironmentAndStatusOrderByCreatedAtDesc(
                userId, environment, TokenStatus.ACTIVE
            ).orElse(null)
            
            existingToken?.startRefresh()
            existingToken?.let { kisTokenRepository.save(it) }
            
            // 3. 새 토큰 발급
            val tokenResponse = kisApiClient.getAccessToken(
                appKey = kisAccount.appKey,
                appSecret = kisAccount.appSecret,
                environment = environment
            )
            
            if (existingToken != null) {
                // 4a. 기존 토큰 갱신
                existingToken.completeRefresh(
                    newAccessToken = tokenResponse.access_token,
                    newExpiresAt = tokenResponse.getExpiresAt()
                )
                val refreshedToken = kisTokenRepository.save(existingToken)
                
                logger.info("Successfully refreshed KIS token for user: $userId")
                return KisTokenDto.from(refreshedToken)
                
            } else {
                // 4b. 새 토큰 생성 (기존 토큰이 없는 경우)
                val newToken = KisToken.create(
                    userId = userId,
                    kisAccountId = kisAccount.id,
                    accessToken = tokenResponse.access_token,
                    environment = environment,
                    expiresAt = tokenResponse.getExpiresAt()
                )
                
                val savedToken = kisTokenRepository.save(newToken)
                
                logger.info("Created new KIS token for user: $userId (no existing token)")
                return KisTokenDto.from(savedToken)
            }
            
        } catch (exception: Exception) {
            logger.error("Failed to refresh KIS token for user: $userId", exception)
            
            // 갱신 실패 시 토큰 상태 업데이트
            kisTokenRepository.findFirstByUserIdAndEnvironmentAndStatusOrderByCreatedAtDesc(
                userId, environment, TokenStatus.REFRESHING
            ).ifPresent { token ->
                token.failRefresh()
                kisTokenRepository.save(token)
            }
            
            throw KisTokenException("토큰 갱신 실패: ${exception.message}", exception)
        }
    }
    
    /**
     * 계정 검증
     * 
     * @param userId 사용자 ID
     * @param request 계정 요청 정보
     * @return 검증 결과
     */
    suspend fun validateKisAccount(userId: Long, request: KisAccountRequest): Boolean {
        logger.info("Validating KIS account for user: $userId, environment: ${request.environment}")
        
        return try {
            val validationResponse = kisApiClient.validateAccount(
                appKey = request.appKey,
                appSecret = request.appSecret,
                environment = request.environment
            )
            
            if (validationResponse.isValid) {
                logger.info("KIS account validation successful for user: $userId")
            } else {
                logger.warn("KIS account validation failed for user: $userId - ${validationResponse.message}")
            }
            
            validationResponse.isValid
            
        } catch (exception: Exception) {
            logger.error("KIS account validation error for user: $userId", exception)
            false
        }
    }
    
    /**
     * 사용자별 환경별 활성 토큰 조회
     */
    @Transactional(readOnly = true)
    fun getActiveToken(userId: Long, environment: KisEnvironment): KisTokenDto? {
        return kisTokenRepository.findFirstByUserIdAndEnvironmentAndStatusOrderByCreatedAtDesc(
            userId, environment, TokenStatus.ACTIVE
        ).map { token ->
            // 토큰 사용 기록
            token.markUsed()
            kisTokenRepository.save(token)
            KisTokenDto.from(token)
        }.orElse(null)
    }
    
    /**
     * 자동 토큰 갱신 스케줄러 (1시간마다 실행)
     */
    @Scheduled(fixedRate = 1, timeUnit = TimeUnit.HOURS)
    fun scheduleTokenRefresh() {
        logger.info("Starting scheduled token refresh")
        
        val now = LocalDateTime.now()
        val tokensToRefresh = kisTokenRepository.findByStatusAndNextRefreshAtBeforeOrderByNextRefreshAtAsc(
            TokenStatus.ACTIVE, now
        )
        
        logger.info("Found ${tokensToRefresh.size} tokens to refresh")
        
        tokensToRefresh.forEach { token ->
            try {
                runBlocking {
                    refreshToken(token.userId, token.environment)
                }
                logger.info("Auto-refreshed token for user: ${token.userId}, environment: ${token.environment}")
                
            } catch (exception: Exception) {
                logger.error("Failed to auto-refresh token for user: ${token.userId}", exception)
            }
        }
        
        // 만료된 토큰 정리
        cleanupExpiredTokens()
    }
    
    /**
     * 만료된 토큰 정리 (배치 작업)
     */
    @Scheduled(fixedRate = 6, timeUnit = TimeUnit.HOURS)
    fun cleanupExpiredTokens() {
        logger.info("Starting expired token cleanup")
        
        val now = LocalDateTime.now()
        
        // 만료된 활성 토큰을 만료 상태로 변경
        val updatedCount = kisTokenRepository.bulkUpdateExpiredTokens(
            currentStatus = TokenStatus.ACTIVE,
            newStatus = TokenStatus.EXPIRED,
            expireTime = now,
            updateTime = now
        )
        
        logger.info("Updated $updatedCount expired tokens")
        
        // 오래된 만료 토큰 삭제 (30일 이후)
        val oldTokens = kisTokenRepository.findByStatusAndExpiresAtBefore(
            TokenStatus.EXPIRED, 
            now.minusDays(30)
        )
        
        if (oldTokens.isNotEmpty()) {
            kisTokenRepository.deleteAll(oldTokens)
            logger.info("Deleted ${oldTokens.size} old expired tokens")
        }
    }
    
    /**
     * 계정 저장 또는 업데이트
     */
    private suspend fun saveOrUpdateKisAccount(userId: Long, request: KisAccountRequest): KisAccount {
        val existingAccount = kisAccountRepository.findByUserIdAndEnvironmentAndAccountNumberAndIsActiveTrue(
            userId, request.environment, request.accountNumber
        )
        
        return if (existingAccount.isPresent) {
            // 기존 계정 업데이트
            val account = existingAccount.get()
            account.appKey = request.appKey
            account.appSecret = request.appSecret
            account.accountAlias = request.accountAlias
            account.markValidated()
            kisAccountRepository.save(account)
        } else {
            // 새 계정 생성
            val newAccount = KisAccount.create(
                userId = userId,
                appKey = request.appKey,
                appSecret = request.appSecret,
                accountNumber = request.accountNumber,
                environment = request.environment,
                accountAlias = request.accountAlias
            )
            kisAccountRepository.save(newAccount)
        }
    }
    
    /**
     * 기존 활성 토큰 비활성화
     */
    private fun deactivateExistingTokens(userId: Long, environment: KisEnvironment) {
        val activeTokens = kisTokenRepository.findByUserIdAndEnvironmentAndStatus(
            userId, environment, TokenStatus.ACTIVE
        )
        
        activeTokens.forEach { token ->
            token.revoke()
        }
        
        if (activeTokens.isNotEmpty()) {
            kisTokenRepository.saveAll(activeTokens)
            logger.debug("Deactivated ${activeTokens.size} existing tokens for user: $userId")
        }
    }
}


/**
 * KIS 토큰 예외 클래스
 */
class KisTokenException(
    message: String,
    cause: Throwable? = null
) : RuntimeException(message, cause)

/**
 * KIS 계정 찾기 실패 예외
 */
class KisAccountNotFoundException(
    message: String
) : RuntimeException(message)